main/simulator.cpp

```diff
@@ -1,9 +1,115 @@
 #include "main/simulator.h"
 
+Simulator::Simulator(float frequency, size_t numOfRaysPerRow, int collectors,
+                     float simulationRadius, std::string_view objPath,
+                     int depth, float sourcePower)
+    : frequency_(frequency), numOfRaysPerRow_(numOfRaysPerRow),
+      manager_(SceneManager(collectors, simulationRadius)), depth_(depth),
+      sourcePower_(sourcePower) {
+  // TODO: no implementation yet
+  // model_ = manager_.getModelTriangles();
+
+  if (!manager_.loadModel(objPath)) {
+    std::stringstream ss;
+    ss << "Invalid model path";
+    throw std::invalid_argument(ss.str());
+  }
+
+  // TODO: remove
+  model_ = manager_.getReferencePlate();
+  collectors_ = manager_.getEnergyCollectors();
+
+  source_ = generators::PointSource(frequency_, numOfRaysPerRow_,
+                                    manager_.getSmapleSize(), &offsetGenerator_,
+                                    simulationRadius, sourcePower);
+}
+
 void Simulator::run() {
-  // TODO: runs simulation
+  for (size_t xIter = 0; xIter < numOfRaysPerRow_; ++xIter) {
```

Comment:
Some comments:

1. size_t
2. I am confused about the naming:
  - (small) x**Iter** suggests this is an iterator, while it is not. How about just "x"
  - numRaysPerRow_ is used in both rows and columns.
3. (serious) numOfRaysPerRow_ seems to be **strictly** tight to the PointSource instance you create in constructor, so having it here seems like a violation of the PointSource API. It seems to be like there is only one possible and correct way of generating the rays from the point source (i.e integer from 0 to N, so that everything is within bounds), so why not just moving it to the source itself and calling the new function here? Note that simulator does not care about the x/y coordinates specifically. For example how about doing this:

```c++
Ray ray;
while (point_source.generateRay(&ray)) {
  rayTrace(ray, depth_);
}
```

More general point: I noticed that you very much like writing code so that it does everything at the same time, and rely on side effects to do the work. For example here, `rayTrace` does 2 things: actually bouncing rays and then distributing the energy to collectors, via a side effect (i.e rayTrace does not return anything so we know something has changed, because the members changed).

This style of coding is really hard to follow, and I think it is much better to:
1. follow SRP -- single responsibility principle (read what this is)
2. strive to avoid side-effects, and make sure to return proper values as results of the function calls

In this case it would be much easier to follow if you wrote the code like this;

```c++
Ray ray;
RayHitData sphereHit;
while (point_source.generateRay(&ray)) {
  if (!rayTrace(ray, depth_, &sphereHit)) {
    continue;
  }

  distributeEnergyFromHit(sphereHit);
}
```

Now you can implement distributeEnergyFromHit however you want, and the rayTrace does not have to know how to distribute rays. This is easier to test as well, because currently when calling rayTrace you need to test both bouncing and energy distribution.

In my approach you can test separately, and a bug in one does not affect the other, and it is much easier to create small self-contained test data for each method.